### PR TITLE
[FX-4608] Migrate Badge to TailwindCSS

### PR DESCRIPTION
**Pull request by ruslan-sed**
[FX-4608]

### Description

Migrate Badge to TailwindCSS.

### How to test

- [Temploy](https://picasso.toptal.net/fx-4608-migrate-badge)
- FIXME: Add the steps describing how to verify your changes

### Development checks

- [x] Add changeset according to [guidelines](https://github.com/toptal/picasso/blob/master/docs/contribution/changeset-guidelines.md) (if needed)
- [x] Read [CONTRIBUTING.md](https://github.com/toptal/picasso/blob/master/CONTRIBUTING.md) and [Component API principles](https://github.com/toptal/picasso/blob/master/docs/contribution/component-api.md)
- [x] Make sure that additions and changes on the design follow [Toptal's BASE design](https://design.toptal.net/), and it's been already discussed with designers at #-base-core
- [x] Annotate all `props` in component with documentation
- [x] Create `examples` for component
- [x] Ensure that deployed demo has expected results and good examples
- [x] Ensure the changed/created components have not caused accessibility issues. [How to use accessibility plugin in storybook](https://github.com/toptal/picasso/blob/master/docs/contribution/accessibility.md).
- [x] Self reviewed
- [x] Covered with tests ([visual tests](https://github.com/toptal/picasso/blob/master/docs/contribution/visual-testing.md) included)

**Breaking change**

- [x] ~codemod is created and showcased in the changeset~
- [x] ~test alpha package of Picasso in StaffPortal~

> All **_development checks_** should be done and set checked to pass the
> **GitHub Bot: TODOLess** action

<details>
<summary>PR commands</summary>
<br />

List of available commands:

- `@toptal-bot run package:alpha-release` - Release alpha version
- `@toptal-anvil ping reviewers` - Ping FX team for review

</details>

<details>
<summary>PR Review Guidelines</summary>
<br />

#### When to approve? ✅

**You are OK** with merging this PR and

1. You have no extra requests.
2. You have optional requests.
   1. Add `nit:` to your comment. (ex. `nit: I'd rename this variable from makeCircle to getCircle`)

#### When to request changes? ❌

**You are not OK** with merging this PR because

1. Something is broken after the changes.
2. Acceptance criteria is not reached.
3. Code is dirty.

#### When to comment (neither ✅ nor ❌)

**You want your comments to be addressed** before merging this PR in cases like:

1. There are leftovers like unnecessary logs, comments, etc.
2. You have an opinionated comment regarding the code that requires a discussion.
3. You have questions.

#### How to handle the comments?

1. An owner of a comment is the only one who can resolve it.
2. An owner of a comment must resolve it when it's addressed.
3. A PR owner must reply with ✅ when a comment is addressed.

</details>

[FX-4608]: https://toptal-core.atlassian.net/browse/FX-4608?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by ruslan-sed** [comment]
@toptal-anvil ping reviewers

File: packages/base/Badge/src/Badge/styles.ts [hunk 1]
```diff
@@ -2,6 +2,8 @@ import { rem } from '@toptal/picasso-shared'
 import type { Theme, StyleRules } from '@material-ui/core/styles'
 import { createStyles } from '@material-ui/core/styles'
 
+import type { SizeType, VariantType } from './Badge'
+
 const getSizeProps = (unitInRem: string): StyleRules[keyof {}] => ({
```
Comment (sashuk) [inline packages/base/Badge/src/Badge/styles.ts:7]: It seems like `getSizeProps` is not used anywhere else, maybe we can delete it, so package no longer relies on `@material-ui`?

Screenshot 2024-05-29 at 12 51 13

File: packages/base/Badge/src/Badge/styles.ts [hunk 2]
```diff
@@ -44,3 +46,14 @@ export default ({ palette, typography }: Theme) =>
...
+
+export const classByVariant: Record<VariantType, string> = {
+  white: 'bg-white text-graphite-700 border-gray-400',
+  red: 'text-white bg-red-500 border-red-500',
+}
+
+export const classBySize: Record<SizeType, string> = {
+  small: 'leading-[10px] rounded-[0.75rem] h-3 min-w-3',
+  medium: 'rounded-[1rem] h-4 min-w-4',
+  large: 'px-[3px] rounded-[1.25rem] h-[1.25rem] min-w-[1.25rem]',
```
[the remark this answers is not in the document]
Comment (TomasSlama) [reply packages/base/Badge/src/Badge/styles.ts:58]: also we can use `rounded-full`

[the remark this answers is not in the document]
Comment (ruslan-sed) [reply packages/base/Badge/src/Badge/styles.ts:58]: @TomasSlama @sashuk 

I remember we were discussing that on migration we should use the same measure units as were used before. That is why I used `rem` here.

[the remark this answers is not in the document]
Comment (ruslan-sed) [reply packages/base/Badge/src/Badge/styles.ts:58]: @TomasSlama @sashuk 

One more thing - we dont have `h-5` class right now. Do you want me to add it?

[the remark this answers is not in the document]
Comment (sashuk) [reply packages/base/Badge/src/Badge/styles.ts:58]: Hm, if we extend `height` prop in configuration, then others from https://tailwindcss.com/docs/height are still available, right?

Screenshot 2024-05-29 at 15 07 09

[the remark this answers is not in the document]
Comment (TomasSlama) [reply packages/base/Badge/src/Badge/styles.ts:58]: width and height is extending spacing, where we don¨t have 5,7,9 🤔 
image

yeah, it makes sense to extend the height and width with missing 5,7,9, WDYT @sashuk

[the remark this answers is not in the document]
Comment (ruslan-sed) [reply packages/base/Badge/src/Badge/styles.ts:58]: We also don't have it here - https://toptal-core.atlassian.net/wiki/spaces/Base/pages/3217031216/Spacing

[the remark this answers is not in the document]
Comment (TomasSlama) [reply packages/base/Badge/src/Badge/styles.ts:58]: yeah, in spacing it is by purpose. For us spacing is margins and paddings and we don't want more values there. The problem is width and height props which we should extend I think.

[the remark this answers is not in the document]
Comment (ruslan-sed) [reply packages/base/Badge/src/Badge/styles.ts:58]: @sashuk our theme, we don't extend it, we replace it. `spacing` list is not in `extend` field. [Here](https://github.com/toptal/picasso/blob/a9768dd9cd4a118a4ae61857124eed0fa84e0090/packages/picasso-tailwind/src/index.js#L22).

[the remark this answers is not in the document]
Comment (sashuk) [reply packages/base/Badge/src/Badge/styles.ts:58]: > we don't extend it, we replace it

Yep, this is what I propose, to keep it the way it works, so we do not allow any custom spacings.

Long story short, let's keep thing they are now to be in sync with BASE design.

[the remark this answers is not in the document]
Comment (sashuk) [reply packages/base/Badge/src/Badge/styles.ts:58]: I have thought that we can reuse `h-5`, but it turned out that it takes changing the way we define spacing, so let's not do this 🙂

File: packages/base/Badge/src/Badge/Badge.tsx [hunk 3]
```diff
@@ -49,25 +45,37 @@ export const Badge = forwardRef<HTMLDivElement, Props>(function Badge(
...
-      classes={{
-        badge: cx(classes.root, classes[variant], classes[size], {
-          [classes.static]: !hasChildren,
-        }),
+      slotProps={{
+        root: {
+          className: twMerge(
+            `inline-flex flex-shrink-0 content-middle flex-nowrap justify-normal 
+            text-[10px] font-semibold leading-3 align-middle text-graphite-700 
+            relative top-0 right-0`,
```
Comment (TomasSlama) [inline packages/base/Badge/src/Badge/Badge.tsx:62]:
```suggestion
            top-0 right-0`,
```
we already have relative set below

File: packages/base/Badge/src/Badge/Badge.tsx [hunk 4]
```diff
@@ -49,25 +45,37 @@ export const Badge = forwardRef<HTMLDivElement, Props>(function Badge(
...
+            hasChildren ? 'relative' : 'static',
+            className
+          ),
+          style: style,
+        },
+        badge: {
+          className: twMerge(
+            `border-solid items-center content-center inline-flex flex-nowrap 
+            justify-center z-[1] border px-[1px] border-gray-400 static`,
```
Comment (TomasSlama) [inline packages/base/Badge/src/Badge/Badge.tsx:71]:
```suggestion
            justify-center z-[1] border px-[1px]`,
```
we already set border color in classbyVariant. Also if we move the `static` to the `hasChildren` condition, we don't need to use `twMerge` here as it is the only thing we merge `absolute -> static`
```suggestion
            justify-center z-[1] border px-[1px] border-gray-400 static`,
```

Comment (TomasSlama) [reply packages/base/Badge/src/Badge/Badge.tsx:71]: ok also `px-[1px]` would need to be moved to `small` and `medium` size. But I would try not to use twMerge for these if it is not necessary

**Review by sashuk — APPROVED** [review]
All comments were addressed or discussed, looks good! Asked about merging to `master` (as we were discussing per-week feature branch) in Slack

**Review by TomasSlama — APPROVED** [review]
all the snapshots seems unrelated to this PR, those are coming from the feature branch, I think we can safely merge

**Pull request merged**